### PR TITLE
Add read_all to std:io::Read to parallel write_all in std::io::Write.

### LINKED ISSUE #23174: Add read_full to std::io::ReadExt
Since the `read` function of `std::io::Read` may read less than the length of the buffer passed to it, adding a function named `read_full` which calls `read` in a loop until the buffer is filled can be useful.

### PR DESCRIPTION
**Pull request by novalis**
Fixes #23174

### CODE REVIEW

**Comment by mahkoh** [comment]
Needs an RFC.

**Comment by novalis** [comment]
RFC: https://github.com/rust-lang/rfcs/pull/980

**Comment by archshift** [comment]
What's the status of this?

**Comment by bombless** [comment]
@archshift, seems that this will get merged once the [RFC](https://github.com/rust-lang/rfcs/pull/980) is accepted.

File: src/libstd/io/error.rs [hunk 1]
```diff
@@ -107,6 +107,14 @@ pub enum ErrorKind {
...
     #[stable(feature = "rust1", since = "1.0.0")]
     WriteZero,
+    /// An error returned when an operation could not be completed because a
+    /// call to `read` returned `Ok(0)`.
+    ///
+    /// This typically means that an operation could only succeed if it read a
+    /// particular number of bytes but only a smaller number of bytes could be
+    /// read.
+    #[stable(feature = "rust1", since = "1.0.0")]
+    ShortRead(usize),
```
Comment (bluss) [inline src/libstd/io/error.rs]: This changes the size of the enum dramatically.

File: src/libstd/io/mod.rs [hunk 2]
```diff
@@ -187,6 +187,32 @@ pub trait Read {
...
+    /// This method will continuously call `read` while there is more data to
+    /// read. This method will not return until the entire buffer has been
+    /// successfully read or an error occurs. The first error generated from
+    /// this method will be returned.
+    ///
+    /// # Errors
+    ///
+    /// This function will return the first error that `read` returns.
+    #[stable(feature = "rust1", since = "1.0.0")]
+    fn read_all(&mut self, mut buf: &mut [u8]) -> Result<()> {
```
Comment (nwin) [inline src/libstd/io/mod.rs]: This should give a warning (#25049). `buf` does not need to be `mut`.

**Comment by alexcrichton** [comment]
I'm going to close this in favor of the [associated RFC](https://github.com/rust-lang/rfcs/pull/980), but we can certainly reopen if it's accepted!

**Pull request closed without merging**